ledgerManager.php

```diff
@@ -0,0 +1,319 @@
+	}
+
+	/**
+	 * This method returns a balance for a given date range, OR sets the value
+	 * of the current balance to the class property $overallBalance
+	 *
+	 * @param string|null $startDate
+	 * @param string|null $endDate
+	 * @return void|float
+	 */
+	protected function calculateBalance($startDate = null, $endDate = null)
+	{
+		$totalCredits = $this->getTotalSumCreditTransactions(true, $startDate, $endDate);
+		$totalDebits = $this->getTotalSumCreditTransactions(false, $startDate, $endDate);
+		$balance = $totalCredits - $totalDebits;
+		if (!$startDate && !$endDate) {
+			$this->overallBalance = $balance;
+		}
+		return $balance;
+	}
+
+	/**
+	 * @param string $startDate
+	 * @param string $endDate
+	 */
+	public function getCreditSumForWindow($startDate, $endDate)
+	{
+		return $this->getTotalSumCreditTransactions(true, $startDate, $endDate);
+	}
+
+	/**
+	 * @param string $startDate
+	 * @param string $endDate
+	 */
+	public function getDebitSumForWindow($startDate, $endDate)
+	{
+		return $this->getTotalSumCreditTransactions(false, $startDate, $endDate);
+	}
+
+	/**
+	 * This method returns a balance for a given date, now() by default
+	 *
+	 * @param string|null $startDate
+	 * @param string|null $endDate
+	 * @return float
+	 */
+	public function getBalance($startDate = null, $endDate = null)
+	{
+		if ($startDate || $endDate) {
+			return $this->calculateBalance($startDate, $endDate);
+		} elseif (!isset($this->overallBalance)) {
+			return $this->calculateBalance();
+		}
+		return $this->overallBalance;
+	}
+
+	/**
+	 * @return float
+	 */
+	public function getUnclearedAmount()
+	{
+		$query = $this->pdo->query("
+			select sum(amount) amount
+			from {$this->table}
+			where credit = false and cleared = false
+		");
+		$row = $query->fetch(PDO::FETCH_ASSOC);
+		return (float)$row['amount'];
+	}
+
+	/**
+	 * @return int
+	 */
+	public function numberOfDaysLeftInPayPeriod()
+	{
+		$todaysDate = (int)date('d');
+		if ($todaysDate <= 15) {
+			return 15 - $todaysDate + 1;
+		} else {
+			$query = $this->pdo->query("
+				select (datediff(last_day(now()), now()) + 1) as daysLeft
+			");
+			$row = $query->fetch(PDO::FETCH_ASSOC);
+			return (int)$row['daysLeft'];
+		}
+	}
+
+	/**
+	 * This window retrieves an array of transactions for a given date range.
+	 * Note: The windowStartDate serves as a cutoff in terms of the returned date
+	 * BUT will not affect the rolling balance which is calculated through the
+	 * beginning of time.  ie. For a database containing transactions going back
+	 * to 2014, a windowStartDate of 2015 would return 2015 through now(), but
+	 * but the transactions in 2014 would setup the balance for 2015.
+	 *
+	 * @param string $windowStartDate
+	 * @param string|null $endDate
+	 * @param float|null $runningBalance
+	 * @return mixed[]
+	 */
+	public function retrieveARangeOfTransactions($windowStartDate, $endDate = null)
+	{
+		$allTransactions = [];
+		$runningBalance = $this->getBalance(null, $endDate);
```

Comment:
the method description should mention this variable as proof that it calculates the balance through the ledger epoch.